src/User/AutoLoginMiddleware.php

```diff
@@ -2,15 +2,76 @@
 
 namespace Yiisoft\Yii\Web\User;
 
+use Psr\Http\Server\RequestHandlerInterface;
+use Psr\Http\Message\ResponseFactoryInterface;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Yiisoft\Auth\IdentityInterface;
+use Yiisoft\Auth\IdentityRepositoryInterface;
+use Yiisoft\Yii\Web\User\User;
+
 /**
  * AutoLoginMiddleware automatically logs user in based on "remember me" cookie
  */
-class AutoLoginMiddleware
+class AutoLoginMiddleware implements MiddlewareInterface
 {
     private User $user;
+    private IdentityRepositoryInterface $identityRepository;
 
-    public function __construct(User $user)
-    {
+    public function __construct(
+        User $user,
+        IdentityRepositoryInterface $identityRepository
+    ) {
         $this->user = $user;
+        $this->identityRepository = $identityRepository;
+    }
+
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $data = $this->getIdentityAndDurationFromCookie($request);
+        if ($this->user->login($data['identity'], $data['duration'])) {
+            try {
+                $response = $handler->handle($request);
+            } catch (\Throwable $e) {
+                throw $e;
+            }
+
+            return $response;
+        } else {
+            throw new \Exception("Error authentication");
+        }
+    }
+
+    /**
+     * Determines if an identity cookie has a valid format and contains a valid auth key.
+     * This method is used when [[enableAutoLogin]] is true.
+     * This method attempts to authenticate a user using the information in the identity cookie.
+     * @param ServerRequestInterface $request Request to handle
+     * @return array|null Returns an array of 'identity' and 'duration' if valid, otherwise null.
```

Comment:
It is bad solution to have nullable array return type

Reply:
What do you propose instead?

Reply:
Use CQRS principle

Reply:
I know what CQRS is, but I never implemented it. Could you show me a little example? I am interested in improving this. Thank you

Reply:
It might looks like the following
```php
public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
{
    if ($this->areCredentialsExists($request)) {
        $credentials = $this->parseCredentials($request);

        if (!$this->user->login($credentials['identity'], $credentials['duration'])) {
            throw new \Exception('Auth error');
        }
    }

    return $handler->handle($request);
}

private function areCredentialsExists(ServerRequestInterface $request): bool
{
    return $this->parseCredentials($request) !== [];
}

private function parseCredentials(ServerRequestInterface $request): array
{
    try {
        $data = json_decode($request->getCookieParams(), true, 512, JSON_THROW_ON_ERROR);
    } catch (\JsonException $e) {
        return [];
    }

    if (!is_array($data) || count($data) === 3) {
        return [];
    }

    [$id, , $duration] = $data;
    $identity = $this->identityRepository->findIdentity($id);
    if ($identity === null) {
        return [];
    }

    return ['identity' => $identity, 'duration' => $duration];
}
```

Reply:
Great, I like this is more readable, I made a mix between this and another proposal